optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/manager/DefaultSolverManager.java

```diff
@@ -0,0 +1,207 @@
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.function.Consumer;
+
+import org.optaplanner.core.api.score.Score;
+import org.optaplanner.core.api.solver.SolverFactory;
+import org.optaplanner.core.api.solver.manager.SolverManager;
+import org.optaplanner.core.api.solver.manager.SolverStatus;
+import org.optaplanner.core.impl.solver.thread.ThreadUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultSolverManager<Solution_> implements SolverManager<Solution_> {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultSolverManager.class);
+
+    private ExecutorService solverExecutorService;
+    private ExecutorService eventHandlerExecutorService;
+    private SolverFactory<Solution_> solverFactory;
+    private ConcurrentMap<Object, SolverTask<Solution_>> problemIdToSolverTaskMap;
+
+    public static <Solution_> SolverManager<Solution_> createFromXmlResource(String solverConfigResource) {
+        Objects.requireNonNull(solverConfigResource);
+        return new DefaultSolverManager<>(solverConfigResource, null, null);
+    }
+
+    public static <Solution_> SolverManager<Solution_> createFromXmlResource(String solverConfigResource, ClassLoader classLoader) {
+        Objects.requireNonNull(solverConfigResource);
+        Objects.requireNonNull(classLoader);
+        return new DefaultSolverManager<>(solverConfigResource, classLoader, null);
+    }
+
+    public static <Solution_> SolverManager<Solution_> createFromXmlResource(String solverConfigResource, ThreadFactory threadFactory) {
+        Objects.requireNonNull(solverConfigResource);
+        Objects.requireNonNull(threadFactory);
+        return new DefaultSolverManager<>(solverConfigResource, null, threadFactory);
+    }
+
+    public static <Solution_> SolverManager<Solution_> createFromXmlResource(String solverConfigResource,
+                                                                             ClassLoader classLoader, ThreadFactory threadFactory) {
+        Objects.requireNonNull(solverConfigResource);
+        Objects.requireNonNull(classLoader);
+        Objects.requireNonNull(threadFactory);
+        return new DefaultSolverManager<>(solverConfigResource, classLoader, threadFactory);
+    }
+
+    private DefaultSolverManager(String solverConfigResource, ClassLoader classLoader, ThreadFactory threadFactory) {
+        if (classLoader != null) {
+            solverFactory = SolverFactory.createFromXmlResource(solverConfigResource, classLoader);
+        } else {
+            solverFactory = SolverFactory.createFromXmlResource(solverConfigResource);
+        }
+
+        problemIdToSolverTaskMap = new ConcurrentHashMap<>();
```

Comment:
Since you're not sizing the `Map`, you can keep the initialization at the declaration site.